building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
++ add_test("test2" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}" "someparam")
+```
+
+- `include` includes a module, some module are builtin in CMake
+- Creates target `test`
+- `make test`, explain output
+- We will learn later (after Christmas) how to work with proper packages for testing (e.g. boost)
+
+### CPack
+
+- We can also package code with cmake (or better: steer from cmake)
+
+`CMakeLists.txt`:
+
+```diff
+include(CPack)
+
+install(TARGETS "${PROJECT_NAME}" DESTINATION bin)
+```
+
+- Goes to `$CMAKE_INSTALL_PREFIX/bin`
+- Run cmake and `make package`, creates `HelloWorld-0.1.1-Linux.tar.gz`
+- We will see later (next week) how to package a `.deb`
+- Set version number
+
+```diff
++ SET(CPACK_PACKAGE_VERSION_MAJOR "1")
++ SET(CPACK_PACKAGE_VERSION_MINOR "2")
++ SET(CPACK_PACKAGE_VERSION_PATCH "3")
+include(CPack)
+```
+
+- Variables must be set before `include(CPack)`
+- `make package` again
+
+### Building libraries
+
+- Let's turn the `sse` function into a library
+
+```diff
++ file(GLOB_RECURSE SRC_FILES CONFIGURE_DEPENDS sse/*.cpp)
+- file(GLOB_RECURSE SRC_FILES CONFIGURE_DEPENDS main.cpp sse/*.cpp)
++ add_library(sse "${SRC_FILES}")
+- add_executable("${PROJECT_NAME}" "${SRC_FILES}")
+```
+
+- Build static lib `libHelloWorld.a`, that's the default
+- Build shared lib by `cmake -DBUILD_SHARED_LIBS=ON ..`, don't change in `CMakeLists.txt`, but stick to standards
+- Define cmake variables with `-D`, we will come back to this later
+- Now, let's use the library
+
+```diff
++ add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- ... gives an error, we also need to link
+
+```diff
++ target_link_libraries("${PROJECT_NAME}" PUBLIC sse)
+```
+
+TODO: explain PRIVATE (only build), PUBLIC (both), INTERFACE (only linking)
+
+### External Dependencies
+
+- preCICE should be used in every code :)
+
+`main.cpp`:
+
+```diff
++ #include "precice/SolverInterface.hpp"
+...
++ sse();
++ std::cout << precice::getVersionInformation() << std::endl;
+```
+
+`CMakeLists.txt`:
+
+```diff
+find_package(precice REQUIRED CONFIG)
+target_link_libraries("${PROJECT_NAME}" PRIVATE precice::precice)
+```
+
+- `find_package` works if dependency is "cmake-ready"
+- Alternative option if lib is in a discoverable location (standard system path or adjusted `LD_LIBRARY_PATH`)
+
+```cmake
+find_library(precice REQUIRED CONFIG)
+target_link_libraries("${PROJECT_NAME}" PRIVATE precice)
+```
+
+- TODO: external dependencies via pkg-config?
+    - @fsimonis: Is `find_package` already using pkg-config? Or is there a real third option?
+    - include(..), pkg-config.search_modules ... link ... cmake searches in pkg-config
+
+### Options and variables
+
+- For good software, a user should not need to touch `CMakeLists.txt`
+- Use variables to talk to cmake
+- Let's try to make preCICE an optional dependency of our code (not really everybody has preCICE installed, sadly)
+
+`CMakeLists.txt`:
+
+```cmake
+if (ENABLE_PRECICE)
```

Comment:
To make this show up in CMake UIs:
```
option(ENABLE_PRECICE ON)
```

Also, you could make precice optional (which I personally don't like)

```cmake
find_library(precice CONFIG)

if (precice_FOUND)
message(STATUS "preCICE found!")
else()
message(STATUS "preCICE sadly not found :'(")
endif()
```

Reply:
You do not like making preCICE optional because you like preCICE or is there a specific reason to this in the sense that it is bad practice or gives problems during the CMake build file generation process?